### PR TITLE
post: rust packages, crates, and modules

### PR DESCRIPTION
**Pull request by roman-kashitsyn**

### CODE REVIEW

File: posts/03-rust-packages-crates-modules.html.pm [hunk 1]
```diff
@@ -0,0 +1,633 @@
...
+  fn get_latest_state(&self) -> State;
+
+  fn commit_state(&self, state: State, version: Version);
+}
+}
+◊figcaption{A generic version of the ◊code{StateManager} trait that does not depend on ◊code{ReplicatedState}.}
+}
+
+◊p{
+  This little trick safes us a lot of compilation time:
```
Comment (dsarlis) [inline posts/03-rust-packages-crates-modules.html.pm:211]:
```suggestion
  This little trick saves us a lot of compilation time:
```

File: posts/03-rust-packages-crates-modules.html.pm [hunk 2]
```diff
@@ -0,0 +1,633 @@
...
+}
+◊ul[#:class "arrows"]{
+◊li{
+  Most Rust libraries do not rely on true global variables.
+  The usual way to pass implicit state around is to use ◊a[#:href "https://doc.rust-lang.org/stable/std/macro.thread_local.html"]{thread-local} state.
+  This becomes a problem when you start spawning new threads: these threads tend to inherit the values of thread locals that you did not expect.
+}
+◊li{
+  Cargo runs tests in parallel by default.
+  If you're not careful with how you're passing loggers between threads, your test output might become an untangeable mess.
```
Comment (dsarlis) [inline posts/03-rust-packages-crates-modules.html.pm:268]:
```suggestion
  If you're not careful with how you're passing loggers between threads, your test output might become an intangible mess.
```
Did you maybe mean intangible here?

File: posts/03-rust-packages-crates-modules.html.pm [hunk 3]
```diff
@@ -0,0 +1,633 @@
...
+  Some people in other programming communities also realized that global loggers are evil.
+  You might enjoy reading ◊a[#:href "https://www.yegor256.com/2019/03/19/logging-without-static-logger.html"]{Logging Without a Static Logger}, for example.
+}
+
+◊advice["dedup-dependencies"]{Deduplicate dependencies.}
+
+◊p{
+  Cargo makes it easy to add dependencies to your code, but it does not provide any tools to consolidate and maintain them in a large workspace.
+  At least until cargo developers implement ◊a[#:href "https://github.com/rust-lang/rfcs/pull/2906"]{RFC 2906}.
+  Until then, every time you bump a version of a dependency, try to do it consistently in all the packages in your workspace.
```
Comment (dsarlis) [inline posts/03-rust-packages-crates-modules.html.pm:322]: Shouldn't we mention that one option is to use `cargo update` that will automatically try to update all dependencies in your `Cargo.lock` file?

**Review by dsarlis — APPROVED** [review]

File: posts/03-rust-packages-crates-modules.html.pm (hunk 3 above)
Comment (roman-kashitsyn) [reply posts/03-rust-packages-crates-modules.html.pm]: Great point, mentioned `cargo updated`.

**Pull request merged**